kotest-assertions/kotest-assertions-core/src/commonMain/kotlin/io/kotest/matchers/concurrent/suspension/concurrent.kt

```diff
@@ -0,0 +1,78 @@
+package io.kotest.matchers.concurrent.suspension
+
+import io.kotest.assertions.assertSoftly
+import io.kotest.assertions.failure
+import io.kotest.common.nonDeterministicTestTimeSource
+import kotlinx.coroutines.TimeoutCancellationException
+import kotlinx.coroutines.withTimeout
+import kotlin.contracts.InvocationKind.EXACTLY_ONCE
+import kotlin.contracts.contract
+import kotlin.time.Duration
+import kotlin.time.measureTime
+import kotlin.time.measureTimedValue
+
+/**
+ * Assert [operation] completes within [duration].
+ *
+ * Note: It does not work well within [assertSoftly].
+ * If used within [assertSoftly] and this assertion fails, any later assertions won't run.
+ */
+suspend fun <A> shouldCompleteWithin(
+   duration: Duration,
+   operation: suspend () -> A,
+): A {
+   contract { callsInPlace(operation, EXACTLY_ONCE) }
+
+   try {
+      return withTimeout(duration) {
+         operation()
+      }
+   } catch (ex: TimeoutCancellationException) {
+      throw failure(
+         "Operation took longer than expected. " +
+            "Expected that operation completed within $duration, but it took longer and was cancelled."
+      )
+   }
+}
+
+/**
+ * Assert [operation] completes within [durationRange].
+ *
+ * Note: It does not work well within [assertSoftly].
+ * If used within [assertSoftly] and this assertion fails, any later assertions won't run.
+ */
+suspend fun <A> shouldCompleteBetween(
+   durationRange: ClosedRange<Duration>,
+   operation: suspend () -> A,
+): A {
+   contract { callsInPlace(operation, EXACTLY_ONCE) }
+
+   val timeSource = nonDeterministicTestTimeSource()
```

Comment:
As I understand it, the Kotest timeout matchers were all using `withTimeout()` before, which responds to virtual time. This would introduce a breaking change to real time for Kotest users, right?

Reply:
Ah good point... I didn't think about that.

So it should just be enough to set `coroutineTestScope = true` in the tests, and use `testTimeSource()` instead of `nonDeterministicTestTimeSource()`?

Reply:
Exactly!